### PR TITLE
tbpore, version 0.1.0

### PR DESCRIPTION
**Pull request by leoisl**
Hello,

this draft PR is just for a WIP review, this is not getting merged now. Please check if you agree with the current changes. This is a small amount of work, just from a single day, but if you spot issues here the best is to solve/change ASAP. Main points so far:

1. H37RV genome and its faidx is explictly added to repo. Do you want to change this approach?
2. Due to python path module issues, I copied what was in `tbpore/tbpore.py` into `tbpore/cli.py`. In summary, `tbpore/tbpore` was first in `sys.path` than `tbpore`, so it would search all the modules (e.g. `from tbpore.constants` in `tbpore/tbpore.py` instead of `tbpore`). By removing `tbpore/tbpore.py` and moving its contents into `tbpore/cli.py`, everything worked in my env. There is probably a better solution than this, allowing us to keep `tbpore/tbpore.py`, but I did not dive too much into it. Do you also have this issue? Should we keep as is described in this PR, or should we roll back into having `tbpore/tbpore.py`?
3. `--threads` and `--mem_mb` parameters added;
4. External tool execution done with class `ExternalTool`;

TODOs:
1. Let user specify log folder for `ExternalTool`s execution;
2. Implement decontamination step (Zam found it better to do all steps, and leave decontamination for later, as it is not that important);
3. Implement the remaining steps;
4. Refactor;
5. Add unit tests;

### CODE REVIEW

File: environment.yaml [hunk 1]
```diff
@@ -12,3 +12,8 @@ dependencies:
   - openssl
   - ripgrep
   - poetry=1.1
+  - mykrobe
```
Comment (mbhall88) [inline environment.yaml:15]: once version 0.11 is out we will need to pin this

File: environment.yaml [hunk 2]
```diff
@@ -12,3 +12,8 @@ dependencies:
...
+  - rasusa
+  - samtools
```
Comment (mbhall88) [inline environment.yaml:17]:
```suggestion
  - samtools=1.13
```

File: environment.yaml [hunk 3]
```diff
@@ -12,3 +12,8 @@ dependencies:
...
+  - gsl
+  - bcftools
```
Comment (mbhall88) [inline environment.yaml:19]:
```suggestion
  - bcftools=1.13
```

File: environment.yaml [hunk 4]
```diff
@@ -12,3 +12,8 @@ dependencies:
   - openssl
   - ripgrep
   - poetry=1.1
+  - mykrobe
+  - rasusa
+  - samtools
+  - gsl
```
Comment (mbhall88) [inline environment.yaml:18]: What is gsl?

File: tbpore/cli.py [hunk 5]
```diff
@@ -29,3 +47,214 @@ def handle_parse_result(self, ctx, opts, args):
...
+    "--threads",
+    help="Number of threads to use in multithreaded tools",
+    type=int,
+    default=1
+)
+@click.option(
+    "-m",
+    "--mem_mb",
+    help="Memory to use in tools that accept a memory limit",
+    type=int,
```
Comment (mbhall88) [inline tbpore/cli.py:109]: Might be more ergonomic to use https://github.com/mbhall88/MemoryUnits. It's a "header-only" python library I wrote.

File: tbpore/cli.py [hunk 6]
```diff
@@ -29,3 +47,214 @@ def handle_parse_result(self, ctx, opts, args):
...
+    # todo: remove contamination
+    # todo: run mykrobe
+    # todo: align decontam to h37rv
+    # todo: bcftools mpileup/call
+    # todo: filter vcf
+    mykrobe = ExternalTool(
+        tool="mykrobe",
+        input=f"-i {infile}",
+        output=f"-o {tmp}/{name}.mykrobe.json",
+        params=f"predict --ont --force -A --debug --format json --min_proportion_expected_depth 0.20 --sample {name} "
```
Comment (mbhall88) [inline tbpore/cli.py:195]: Don't use the `--ont` flag until after the release of mykrobe v0.11. Use https://github.com/mbhall88/head_to_head_pipeline/blob/d7b30eda3cca5fbaf2dc22dbcacd79f7c3a876c9/analysis/resistance_prediction/workflow/rules/mykrobe.smk#L18-L23

See https://github.com/mbhall88/tbpore/issues/2#issue-1196635228

Comment (mbhall88) [reply tbpore/cli.py:195]: I think adding a CLI param for whether to use `-A` is wise. I think it should be off by default. It tells mykrobe to output details for all (~66k) mutations. As such, the JSON file is ~92MB. Also don't think we need `--debug`

File: tbpore/cli.py [hunk 7]
```diff
@@ -29,3 +47,214 @@ def handle_parse_result(self, ctx, opts, args):
...
+
+    sorted_sam_file = f"{tmp}/{name}.subsampled.sorted.sam"
+    samtools_sort = ExternalTool(
+        tool="samtools",
+        input=sam_file,
+        output=f"-o {sorted_sam_file}",
+        params=f"sort -@ {threads}"
+    )
+
+    pileup_file = f"{tmp}/{name}.subsampled.pileup.bcf"
```
Comment (mbhall88) [inline tbpore/cli.py:223]: we're not going to keep the pileup file, so you can use VCF, or even stream it directly into `call` if easier

File: tbpore/cli.py [hunk 8]
```diff
@@ -29,3 +47,214 @@ def handle_parse_result(self, ctx, opts, args):
...
+
+    snps_file = f"{tmp}/{name}.subsampled.snps.bcf"
+    bcftools_call = ExternalTool(
+        tool="bcftools",
+        input=pileup_file,
+        output=f"-o {snps_file}",
+        params=f"call --ploidy 1 -O b -V indels -m --threads {threads}"
+    )
+
+    tools_to_run = [mykrobe, rasusa, minimap, samtools_sort, bcftools_mpileup, bcftools_call]
```
Comment (mbhall88) [inline tbpore/cli.py:240]: This is super neat!!

File: tbpore/external_tools.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        self.command: List[str] = self._build_command(tool, input, output, params)
+        os.makedirs(logdir, exist_ok=True)
+        command_hash = hashlib.sha256(" ".join(self.command).encode("utf-8")).hexdigest()
+        logfile_prefix: Path = logdir / f"{tool}_{command_hash}"
+        self.out_log = f"{logfile_prefix}.out"
+        self.err_log = f"{logfile_prefix}.err"
+
+    @property
+    def command_as_str(self) -> str:
+        return " ".join(self.command)
```
Comment (mbhall88) [inline tbpore/external_tools.py:21]: I wonder if https://docs.python.org/3.8/library/shlex.html#shlex.join might be safer?

**Review by mbhall88 — CHANGES_REQUESTED** [review]
I wonder if the reason the tbpore file didn't work is I accidently forgot to change this after copying it from snakefmt https://github.com/mbhall88/tbpore/blob/603ae00020b797635ca340eae0e0172a5601db2b/pyproject.toml#L13

It *should* work, as it was working for me fine and this is the exact same file structure I use in snakefmt. Are you running tbpore in a poetry env? i.e. `poetry run tbpore -h`

File: tbpore/cli.py (hunk 5 above)
Comment (mbhall88) [reply tbpore/cli.py:109]: Actually, this is probably not worth the effort. Ignore this

**Comment by leoisl** [comment]
> I wonder if the reason the tbpore file didn't work is I accidently forgot to change this after copying it from snakefmt
> 
> https://github.com/mbhall88/tbpore/blob/603ae00020b797635ca340eae0e0172a5601db2b/pyproject.toml#L13
> 
> It _should_ work, as it was working for me fine and this is the exact same file structure I use in snakefmt. Are you running tbpore in a poetry env? i.e. `poetry run tbpore -h`

That is probably my fault as I am not experienced with poetry. I tried to install the packages with `poetry install` (as like in `pip install`) and was trying to run as `PYTHONPATH="." python tbpore/__main__.py`. In the end, I made the aforementioned changes, and extracted a `requirements.txt` from poetry and pip installed them, and it is working now. Next time will try `poetry run tbpore -h`, but will leave tooling as last step now. BTW do you know how the end user will install `tbpore` actually? I did not search much, but I haven't had much success on seamlessly combining conda+poetry or conda+pip. I am thinking about a 2-step approach where the user creates the conda env, and then installs `tbpore` inside the conda env. I am guessing `tbpore` will be deployed to Pypi, giving how easy it is to publish to Pypi if poetry is set up?

File: environment.yaml (hunk 4 above)
Comment (leoisl) [reply environment.yaml:18]: it is a dependency of `bcftools` that I have no idea why it was not getting installed. When running `bcftools` I was getting `bcftools: error while loading shared libraries: libgsl.so.25: cannot open shared object file: No such file or directory` . I tried conda uninstalling, installing again, reinstalling, etc, all with the same error. What fixed for me was explicitly installing `gsl`. Once we are done with the code, I will review the tooling and check if this line is still necessary or superfluous (I'd guess superfluous, as `gsl` has to be a dependency of `bcftools` in `conda`, but who knows...)

File: tbpore/cli.py (hunk 5 above)
Comment (leoisl) [reply tbpore/cli.py:109]: I will actually remove this argument I think, as it is just `mykrobe` who uses it. We can put a fix 4 or 8 GB value. I am also thinking of having a `config.yaml` file that will configure these magical numbers (and lots of magical parameters in the code)

File: tbpore/external_tools.py (hunk 9 above)
Comment (leoisl) [reply tbpore/external_tools.py:21]: Nice spot, will do that, even though `command_as_str` is never executed, is just printed to log. So it should be safe anyway, but I think `shlex.join` will make it even safer

Comment (leoisl) [reply tbpore/external_tools.py]: Done in https://github.com/mbhall88/tbpore/pull/8/commits/39d4cb4c8e226f14e044d8d5a96650d2aabc522f

File: environment.yaml (hunk 1 above)
Comment (leoisl) [reply environment.yaml]: For now we have `0.10` fixed: https://github.com/mbhall88/tbpore/pull/8/commits/b6a5797af99166b8194d3d790ea8ee17ad0af3d4

File: environment.yaml (hunk 2 above)
Comment (leoisl) [reply environment.yaml]: https://github.com/mbhall88/tbpore/pull/8/commits/b6a5797af99166b8194d3d790ea8ee17ad0af3d4

File: environment.yaml (hunk 3 above)
Comment (leoisl) [reply environment.yaml]: https://github.com/mbhall88/tbpore/pull/8/commits/b6a5797af99166b8194d3d790ea8ee17ad0af3d4

File: tbpore/cli.py (hunk 6 above)
Comment (leoisl) [reply tbpore/cli.py]: Done in https://github.com/mbhall88/tbpore/pull/8/commits/aeb1de49fdb00ccef3955f85c75ea5e9ce90b9dc and https://github.com/mbhall88/tbpore/pull/8/commits/b6016a5ff925ef1f64531bf0ff283787dbfe804e

File: tbpore/cli.py (hunk 7 above)
Comment (leoisl) [reply tbpore/cli.py:228]: I think the `ExternalTool` class does not support streaming, is it ok to leave like this?

File: tbpore/cli.py (hunk 8 above)
Comment (leoisl) [reply tbpore/cli.py]: thanks!

**Comment by leoisl** [comment]
Summary of updates from today's commits, excluding those addressing previous comments:

* [Binary skeletons are stored in a `.cache` dir in the repo root ](https://github.com/mbhall88/tbpore/pull/8/commits/d9a0c027c5d4cc7294d803d49fcedde300811475)
* [H37RV mask added](https://github.com/mbhall88/tbpore/pull/8/commits/41d51776d72c26828198ad581c003bd4765056cf)
* [All remaining steps finished](https://github.com/mbhall88/tbpore/pull/8/commits/cd74db904e323cad6510a8c64e1942993852aecc)
* External tools parameters that might (rarely) change, but that are not changeable through `tbpore` args are now listed in a `.config.yaml` file, making them more configurable: https://github.com/mbhall88/tbpore/pull/8/commits/6703cfb0ef007b52dd37a69745adf9a8d3a1ca42 and https://github.com/mbhall88/tbpore/pull/8/commits/b7aa4cfcac2c426a986f05077e9e323f8a0695be
* [`--logdir` parameter added](https://github.com/mbhall88/tbpore/pull/8/commits/27f331a2b90d7efac1c52f2095dc02742a15e870)
* [Mocking the execution of external tools, making tests better](https://github.com/mbhall88/tbpore/pull/8/commits/2d94f3107633ceb22e73533d56cdf34d5f53b06c)
* [Added a test to check all tools are called in the correct order and with the correct arguments](https://github.com/mbhall88/tbpore/pull/8/commits/bc653db145f09d1ddd9aecb7b87ba235d95adb35)
* General refactoring

TODO:
* Add more unit tests;
* Refactor tooling;
* Lint/format;

File: environment.yaml (hunk 4 above)
Comment (mbhall88) [reply environment.yaml:18]: If you pin bcftools to 1.13 this should go away maybe? This sounds weird and buggy

File: tbpore/cli.py (hunk 5 above)
Comment (mbhall88) [reply tbpore/cli.py]: fine by me

File: tbpore/cli.py (hunk 7 above)
Comment (mbhall88) [reply tbpore/cli.py:243]: sure. you can save a tiny bit of time by not using bcf though? can just use vcf. anyway, this is minor

**Comment by mbhall88** [comment]
> That is probably my fault as I am not experienced with poetry. I tried to install the packages with poetry install (as like in pip install) and was trying to run as PYTHONPATH="." python tbpore/__main__.py. In the end, I made the aforementioned changes, and extracted a requirements.txt from poetry and pip installed them, and it is working now. Next time will try poetry run tbpore -h, but will leave tooling as last step now. BTW do you know how the end user will install tbpore actually? I did not search much, but I haven't had much success on seamlessly combining conda+poetry or conda+pip. I am thinking about a 2-step approach where the user creates the conda env, and then installs tbpore inside the conda env. I am guessing tbpore will be deployed to Pypi, giving how easy it is to publish to Pypi if poetry is set up?

The tooling *was* setup already though. It is always easier to set that up at the start. The way I do it is install poetry inside the conda env. Then whenever I want to run tbpore use `poetry run tbpore ...`. You will need to change the leftover `snakefmt` in pyproject.toml for this to work (just checked locally and that was why it wasn't working for you before with the main function in tbpore.py).

I was planning to distribute tbpore on pypi (using poetry) and bioconda. then the installation process is just `conda install tbpore` and you get all the external dependencies.

File: .config.yaml [hunk 10]
```diff
@@ -0,0 +1,42 @@
+# tbpore external tools parameters config file
+# DO NOT EDIT - parameters that should be changed by users are listed in the program options
+# these are parameters that are passive to be changed, so they are listed here
+
+mykrobe:
+  predict:
+    params: "-e 0.08 --ploidy haploid --force --format json --min_proportion_expected_depth 0.20 --species tb -m 8192MB"
```
Comment (mbhall88) [inline .config.yaml:7]: BTW the `--force` flag will override the skeletons, so maybe you can remove that if we are caching them now.
Also, memory of 2GB is more than sufficient

File: environment.yaml [hunk 11]
```diff
@@ -12,3 +12,9 @@ dependencies:
   - openssl
   - ripgrep
   - poetry=1.1
+  - mykrobe=0.10
+  - rasusa
+  - samtools=1.13
+  - gsl
+  - bcftools=1.13
+  - cyvcf2=0.30
```
Comment (mbhall88) [inline environment.yaml:20]: cyvcf2 should be a dependency in pyproject.toml, not in the conda env

File: pyproject.toml [hunk 12]
```diff
@@ -23,6 +23,7 @@ isort = "^5.10.1"
 flake8 = "^4.0.1"
 pytest = "^6.2.5"
 pytest-cov = "^3.0.0"
+pyyaml = "^6.0"
```
Comment (mbhall88) [inline pyproject.toml:26]: if this is for reading in the config, wouldn't this be a dependency, not a dev-dependency?

**Review by mbhall88 — CHANGES_REQUESTED** [review]

File: .config.yaml (hunk 10 above)
Comment (leoisl) [reply .config.yaml]: Done in https://github.com/mbhall88/tbpore/pull/8/commits/f0bdc91b809bc2805ec50758dd7c7d8ddeedb04f

File: environment.yaml (hunk 11 above)
Comment (leoisl) [reply environment.yaml]: Done in https://github.com/mbhall88/tbpore/pull/8/commits/4d67fb54d1fc168f30323bca1474b8d5d56473d1

File: pyproject.toml (hunk 12 above)
Comment (leoisl) [reply pyproject.toml]: Done in https://github.com/mbhall88/tbpore/pull/8/commits/9ed4fd61dab3335ba4df6a86f95402a9a59f8651

File: environment.yaml (hunk 4 above)
Comment (leoisl) [reply environment.yaml]: It did, thanks! Done in https://github.com/mbhall88/tbpore/pull/8/commits/d4578bbe9214dd8b9379f269e4739575ba1b62ed

**Comment by leoisl** [comment]
> I was planning to distribute tbpore on pypi (using poetry) and bioconda. then the installation process is just `conda install tbpore` and you get all the external dependencies.

Perfect for me.

Re: tooling, poetry and `tbpore/tbpore.py`: sorry, old c++ grandpa can't keep up with these amazing modern tooling. `poetry` running just fine, and I [updated it](https://github.com/mbhall88/tbpore/pull/8/commits/d49f4339dd36bae1313a39d2deb18a4ed9baf26a) and [`tbpore/tbpore.py` is back](https://github.com/mbhall88/tbpore/pull/8/commits/4c4c81fccd75c01e80144aaa07f4da58f99f2dfb), so all fine I think tooling-wise

File: tbpore/cli.py (hunk 7 above)
Comment (leoisl) [reply tbpore/cli.py]: Done in https://github.com/mbhall88/tbpore/pull/8/commits/8c3eb8afd323d12afa64c392903a19e762b2f827

**Comment by leoisl** [comment]
Summary of updates from today's commits, excluding those addressing previous comments or issues:

* [Removed --logdir parameter](https://github.com/mbhall88/tbpore/pull/8/commits/36dd798fef1e55a0c8d1c56850b3d57f2ebb2e1c) - I favoured to create a `logs` dir inside `outdir` so that the number of parameters is reduced;
* [Defaulting `--outdir` from `.` to `tbpore_out`](https://github.com/mbhall88/tbpore/pull/8/commits/b9a3396685c8987383186ac3d1b7cb5fb09c0e67) - as such is easier to see which output files were created;
* Finished unit tests (see above for coverage);
* Sorted out tooling (was fine since the start, I was failing to use it well);
* Formatting done;

**Comment by leoisl** [comment]
I think this PR is ready for review. Closes #2, #3, #4, #5, #6, #7, #9, #10. Still open: #11, #12, #13, #14

File: environment.yaml [hunk 13]
```diff
@@ -12,3 +12,7 @@ dependencies:
   - openssl
   - ripgrep
   - poetry=1.1
+  - mykrobe=0.10
```
Comment (mbhall88) [inline environment.yaml:15]: 0.11 was just released, but I am just trying to update the conda recipe to have the correct dependencies

File: pyproject.toml [hunk 14]
```diff
@@ -10,12 +10,14 @@ repository = "https://github.com/mbhall88/tbpore"
 keywords = ["tuberculosis", "nanopore", "diagnostics", "genomics", "variant-calling", "resistance-prediction"]
 
 [tool.poetry.scripts]
-snakefmt = 'tbpore.tbpore:main'
+tbpore = 'tbpore.tbpore:main'
 
 [tool.poetry.dependencies]
 python = ">=3.8,<4.0"
-click = "^8.0.3"
+click = "8.0.3"
```
Comment (mbhall88) [inline pyproject.toml:17]: why did you remove the `^`?

Comment (leoisl) [reply pyproject.toml:17]: I didn't find what `^` mean, but I think means install a higher version? Anyway, with `^` we were installing `click` at a higher version than `8.0.3` (don't remember exactly which one). The [`click` version was incompatible with our `black` version](https://github.com/psf/black/issues/2964), so [I fixed `click` to `8.0.3`](https://github.com/mbhall88/tbpore/pull/8/commits/684912e2adf10773df4abce008e48c9550527c6a), and things worked. If you prefer, we can instead force `black` version to be `>=22.3.0`, which has a fix for this issue, and the `click` can be any version. I personally prefer to fix the version of every dependency to a specific version, as updates can break tools/other dependencies, but I will leave the choice to you

**Comment by leoisl** [comment]
I had to create this [dev branch](https://github.com/mbhall88/tbpore/tree/dev) and change the base branch of this PR from `main` to `dev` because [the latest commit to the main branch](https://github.com/mbhall88/tbpore/commit/2db43e05b2bf76cedef11ec4a8c2915b67621016) creates a conflict with this PR, which causes github actions to not trigger CI

**Comment by leoisl** [comment]
All builds now work, see https://github.com/mbhall88/tbpore/issues/12#issuecomment-1109568596

File: pyproject.toml (hunk 14 above)
Comment (mbhall88) [reply pyproject.toml:17]: https://python-poetry.org/docs/dependency-specification/#caret-requirements

I think leave click with the caret and remove and version pinning for black.

Comment (leoisl) [reply pyproject.toml]: Done in https://github.com/mbhall88/tbpore/pull/8/commits/78619f6f87562c7d159a0420c42b6cc355c8a8e5 . Is it fine keep `black = ">=22.3.0"` to ensure it is compatible with recent `click` versions?

File: environment.yaml (hunk 13 above)
Comment (mbhall88) [reply environment.yaml:15]: ok, bioconda release of v0.11.0 should be available now

File: README.md [hunk 15]
```diff
@@ -9,4 +9,62 @@
...
+
+### Walkthrough
+
+```
+git clone https://github.com/mbhall88/tbpore
+cd tbpore
+conda env create -f environment.yaml && conda activate tbpore  # install dependencies
+just install  # install tbpore
+just check  # checks installation is fine
+scripts/run_sample_example.sh  # if you want to run tbpore in a sample example
```
Comment (mbhall88) [inline README.md:34]:
```suggestion
# if you want to run tbpore on an example isolate
scripts/run_sample_example.sh
```

File: scripts/run_sample_example.sh [hunk 16]
```diff
@@ -0,0 +1,9 @@
+#!/bin/env sh
```
Comment (mbhall88) [inline scripts/run_sample_example.sh:1]:
```suggestion
#!/usr/bin/env sh
```

File: tbpore/tbpore.py [hunk 17]
```diff
@@ -111,35 +195,113 @@ def main(
...
+    snps_file = f"{tmp}/{name}.subsampled.snps.vcf"
+    bcftools_call = ExternalTool(
+        tool="bcftools",
+        input=pileup_file,
+        output=f"-o {snps_file}",
+        params=f"call --threads {threads} {config['bcftools']['call']['params']}",
+        logdir=logdir,
+    )
+
+    filtered_snps_file = f"{outdir}/{name}.subsampled.snps.filtered.vcf"
```
Comment (mbhall88) [inline tbpore/tbpore.py:259]:
```suggestion
    filtered_snps_file = f"{outdir}/{name}.subsampled.snps.filtered.bcf"
```

Probably better to make this a BCF file to save disk space

File: tbpore/utils.py [hunk 18]
```diff
@@ -13,8 +13,8 @@ def is_fastq(p: PathLike) -> bool:
     return Path(p).is_file() and bool(FASTQ_REGEX.search(str(p)))
 
 
-def find_fastq_files(src: Path, recursive: bool = False) -> List[Path]:
-    return [p for p in glob.iglob(f"{src}/**", recursive=recursive) if is_fastq(p)]
+def find_fastq_files(src: Path, recursive: bool = False) -> Set[str]:
+    return set(p for p in glob.iglob(f"{src}/**", recursive=recursive) if is_fastq(p))
```
Comment (mbhall88) [inline tbpore/utils.py:17]: Am I missing something? Where is the import of `glob`? Given `src` is a `Path` might [`rglob`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.rglob) be a more convenient option?

**Review by mbhall88 — APPROVED** [review]
Just a few comments. Feel free to do as you wish with them, but all good to merge when you want.

File: environment.yaml (hunk 13 above)
Comment (leoisl) [reply environment.yaml]: Done in https://github.com/mbhall88/tbpore/pull/8/commits/10b410e9f9e6f865507f1c9d70e17d4418602461

**Comment by leoisl** [comment]
See https://github.com/mbhall88/tbpore/issues/16#issuecomment-1115986893 for a detailed comparison between H2H and `tbpore` (at the tip of this PR) results. If they look ok for you, this PR can then be merged
